source/jupyter/wire/kernel.d

```diff
@@ -167,11 +175,77 @@ struct Kernel(Backend) if(isBackend!Backend) {
             version(JupyterLogVerbose) log("Told by the FE to execute code");
             handleExecuteRequest(requestMessage);
             return;
+
+        case "comm_open":
+            version(JupyterLogVerbose) log("Told by the FE to open a comm");
+            handleCommOpen(requestMessage);
+            return;
+
+        case "comm_msg":
+            version(JupyterLogVerbose) log("Received a comm msg from the FE");
+            handleCommMessage(requestMessage);
+            return;
+
+        case "comm_close":
+            version(JupyterLogVerbose) log("Told by the FE to close a comm");
+            handleCommClose(requestMessage);
+            return;
         }
 
         assert(0);
     }
 
+    void handleCommOpen(Message requestMessage) {
+        import jupyter.wire.message: commCloseMessage;
+        import std.traits: hasMember;
+
+        bool success = false;
+        scope(exit) if (!success) {
+                auto msg = commCloseMessage(requestMessage);
+                sockets.send(sockets.ioPub, msg);
+            }
+
+        static if (hasMember!(typeof(backend), "commOpen")) {
+            scope sender = (Message msg){
+                    msg.parentHeader = requestMessage.header;
+                    sockets.send(sockets.ioPub, msg);
+                };
+            success = backend.commOpen(requestMessage.content["comm_id"].str,
+                                       requestMessage.content["target_name"].str,
+                                       requestMessage.metadata["version"].str,
+                                       requestMessage.content["data"],
```

Comment:
This means passing 3 strings in a row, which makes for a confusing API. Debatable whether this is important since users aren't calling this function directly.

Reply:
Widgets can be created either on the frontend or the backend. When either side creates one it sends a `comm_open` message to the other. Widgets will need to be implemented in the backend so so we need a way to forward those messages.

I agree on the 3 strings in a row. Will create a struct and pass that instead.